fix: json2satrec - parse EPOCH ending with Z

timotejroiko opened the pull request:
This PR fixes an issue where `json2satrec` is not able to parse EPOCH strings that end with Z.
While most OMM providers tend to omit the trailing Z from their EPOCHs, the OMM standard does allow a trailing Z and most dates produces by JS software do end with Z by default (ie: `new Date().toISOString()`).

This PR also marks `REV_AT_EPOCH` as optional since its not required by the spec.

`OMMJsonObject` type changes:

* make REV_AT_EPOCH optional (not required by the spec and not used in any calculations)

`json2satrec` function changes:

* correctly parse EPOCH that ends with "Z"

<hr>

<s>
This PR aims to make `json2satrec` less strict about the properties it requires to function and also compatible with alternative epoch formats.

type changes:

* make OBJECT_NAME optional (not used in any calculations)
* make OBJECT_ID optional (not used in any calculations)
* make REV_AT_EPOCH optional (not used in any calculations)

function changes:

* make EPOCH also accept milliseconds and ISO string that already ends with "Z" (actual type unchanged)
</s>

timotejroiko committed:
make json2satrec less strict

1valdis commented:
Hi, thanks for the PR! I assume a need to process an OMM output of some software, without having to pre-process the OMM before passing it to the library, would be the rationale behind the PR. So out of curiosity: what is the use case? Which software outputs this sort of OMM?

Now on the question of `EPOCH` field accepting milliseconds. Citing OMM standard, section 6.5.9, which is referenced from the `EPOCH` field format requirements:
> 6.5.9 In value fields that represent a time tag or epoch, times shall be given in one of the following two formats:
> YYYY-MM-DDThh:mm:ss[.d→d][Z]
> or
> YYYY-DDDThh:mm:ss[.d→d][Z]
> where ‘YYYY’ is the year, ‘MM’ is the two-digit month, ‘DD’ is the two-digit day, ‘DDD’ is the three-digit day of year, ‘T’ is constant, ‘hh:mm:ss[.d→d]’ is the time in hours, minutes seconds, and optional fractional seconds; ‘Z’ is an optional time code terminator (the only permitted value is ‘Z’ for Zulu, i.e., UTC). As many ‘d’ characters to the right of the period as required may be used to obtain the required precision, up to the maximum allowed for a fixed point number. All fields shall have leading zeros.

Since milliseconds don't conform to OMM standard (unless I'm missing something) and aren't used in major data providers (Celestrak, Space-Track), I'm currently not agreeing on including this change. The trailing `Z` fix though is on point!

`OBJECT_NAME` and `OBJECT_ID` are obligatory by the OMM standard too, so there should be a good reason to make them optional.

timotejroiko commented:
Hi, thanks for the reply.

I am aware that milliseconds is not part of the standard, hence i did not change the object type, but rather chose to make the function itself accept it.

The rationale behind it is mostly a personal nitpick for data efficiency. My use case is that i store local copies of OMM data in more memory-efficient formats to conserve space, including removing redundant/unused data and converting the ISO string to milliseconds.

I would argue that it would be a advantageous to have the OMMJsonObjectV3 typing conform to the standards so it can be used for that reason where needed (data validation, etc...) but to have the actual functionality of the program only require what it actually uses within its calculations, to avoid redundant data manipulation.

The idea for this PR came to mind when i found myself having to manually parse the EPOCH back to ISO and then manually remove the Z from the end (since JS's `Date.toISOString()` method already includes the Z) only for the program to add the Z back to it and convert it back, so i asked myself why are we doing this unnecessary round trip?

As for OBJECT_NAME and OBJECT_ID they are not the only properties that are obligatory by the standard but were made optional in this program, one would think that perhaps them not being actually used anywhere within the calculations would be a good enough reason.

But i do understand that perhaps such changes would be out of scope for this project, so let me know what you don't agree with and I'll have it removed from this PR, no worries.

1valdis commented:
Thanks for the clarifications.

The most significant pain point in implementing OMM was that both major OMM providers, Celestrak and Space-Track, don't fully conform to the OMM format. With this library, I don't want to further dismiss OMM standard requirements and encourage even more deviations. This meant that the library either had to fully conform to the OMM and force users to pre-process virtually all their OMM data (which would make it unusable), or, the option that I chose, is the library has the OMM type that is relaxed only to a point that allows major providers data to be processed - but not further down. That said, on the changes proposed by the PR:

1. Since both providers do provide `OBJECT_NAME` and `OBJECT_ID`, and they are obligatory in the format, I think this should stay this way, otherwise that would mean a deviation from the strictest possible format given by the providers.
2. The rationale of no deviations also applies to the milliseconds feature. I don't think this should be relaxed. Some databases have corresponding types for memory-efficient and ISO8601-conforming storage, for example Postgres with `timestamptz` taking just 8 bytes. Not to mention that JSON is the second most verbose format of OMM anyway, behind only XML.
3. The fix of parsing the date value with `Z` is much appreciated. I was implementing parsing while looking at the data from the providers and they didn't have the trailing `Z`, but the standard permits it, which means that should get fixed. In hindsight, the `YYYY-DDDThh:mm:ss` format should also be properly parsed, but I have yet to encounter any software or source that uses that at all, so this can probably be postponed until someone actually needs that🤔

timotejroiko committed:
changes

timotejroiko commented:
Hi, your concerns are valid and understandable.

I have updated the PR to remove the other changes, and keep only the Z fix and the REV_AT_EPOCH change (since its not required in the spec, i believe this should be fine).

i have also updated the title and first post to reflect the changes.

1valdis commented:
Thank you very much! I'll release this later today.

1valdis commented:
Released as `6.0.1`.
